mixxx::Logger

uklotzde opened the pull request:
A simple logger that provides the ability to consistently add context information like a class name to log messages. It also helps to transparently use the log levels _Info_ and _Fatal_ in Qt4 code, which are currently not or only partially supported.

As an example I've replaced all Qt logging primitives with the new mixxx::Logger in SoundSources.

uklotzde committed:
Use the new mixxx::Logger

uklotzde committed:
Add a simple mixxx::Logger

uklotzde committed:
Remove support for log level Fatal from mixxx::Logger

...no QDebug version of qFatal() available, neither for Qt4 nor Qt5.

uklotzde committed:
Fix unresolved external symbol errors for SoundSource plugins

uklotzde committed:
Fix a missing #include on Windows

uklotzde commented:
OS X built timed out, although it previously finished without any erros.

uklotzde committed:
Fix missing database connection reference

...otherwise the default connection is chosen.

uklotzde committed:
Add some convenience functions to mixxx::Logger

uklotzde committed:
Log available SoundSource providers on level Info

daschuer commented on src/util/logging.cpp:
```diff
@@ -241,4 +241,31 @@ void Logging::shutdown() {
+
+} // anonymous namespace
+
+Logger::Logger(const char* logContext)
```
Shouldn't the file be called "logger.cpp"?

daschuer commented on src/util/logging.cpp:
```diff
@@ -241,4 +241,31 @@ void Logging::shutdown() {
+} // anonymous namespace
+
+Logger::Logger(const char* logContext)
+    : m_preambleChars(preambleString(logContext).toLocal8Bit()) {
```
Can't we return the local 8 bit QByteArray from "preambleString()"? 
I think we can get around the intermediate QString by just append QByteArrays.

daschuer commented on src/util/logging.cpp:
```diff
@@ -241,4 +241,31 @@ void Logging::shutdown() {
+namespace {
+
+inline QString preambleString(const char* logContext) {
+    if ((logContext == nullptr) || (strlen(logContext) == 0)) {
```
logContext is typical a string literal. Can we use QLatin1String interface instead?

daschuer commented:
Thank you for the nice Idea and realization. I have just added some optimization nits.

uklotzde committed:
Move logger into separate .h/.cpp files

uklotzde committed:
Optimize string concatenation when creating loggers

uklotzde committed:
Restrict log context from QString to QLatin1String

uklotzde committed:
Fix missing member initialization in AutoDJCratesDAO

...it simply seems to use the default database connection!? More fixes
towards safe, multi-threaded datatabase access will follow soon ;)

uklotzde commented:
AppVeyor upload failure for Windows x64 build:
```
[00:36:44] Exception calling "UploadFile" with "2" argument(s): "The remote server returned an error: (500) Internal Server Error."
[00:36:44] At line:1 char:88
[00:36:44] + ... /ci.appveyor.com/api/testresults/junit/$($env:APPVEYOR_JOB_ID)", (Res ...
[00:36:44] +                                              ~~~~~~~~~~~~~~~~~~~~
[00:36:44]     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
[00:36:44]     + FullyQualifiedErrorId : WebException
[00:36:44]  
[00:36:44] Command executed with exception: Exception calling "UploadFile" with "2" argument(s): "The remote server returned an error: (500) Internal Server Error."
[00:36:44] echo "*** FAILURE ***"
```

uklotzde committed:
Add log level "trace" for profiling messages

uklotzde committed:
Reduce precision of performance timer traces

uklotzde committed:
Add a function for profiling performance to the logger

uklotzde committed:
Prettify performance trace logs

daschuer commented on src/util/logger.cpp:
```diff
@@ -0,0 +1,36 @@
+namespace {
+
+const char* const kLogPreambleSuffix = " -";
+const std::size_t kLogPreambleSuffixLen = std::strlen(kLogPreambleSuffix);
```
if you use here a QLatin1String, the size() member can be used without cost. Alternative, a constexp may work.

daschuer commented on src/util/logger.cpp:
```diff
@@ -0,0 +1,36 @@
+inline QByteArray preambleChars(const char* logContext) {
+    QByteArray preamble;
+    if (logContext != nullptr) {
+        std::size_t logContextLen = std::strlen(logContext);
```
Did you consider to use here QLatin1String as well, to get around searching for \0?

uklotzde committed:
Use QLatin1String to avoid null check

uklotzde committed:
Restrict usage of C string functions to Qt4

uklotzde committed:
Fix Qt5 build

daschuer commented:
Perfect, sorry for putting too much extra work on you, I was not aware about the QT4/5 issues.
LGTM

Be-ing commented:
It would be nice if there was a way, such as a command line option, to hide debugging output from parts of Mixxx from the console output. Output from scrolling in the library, loading tracks, and analyzing tracks is distracting from debugging other issues.

uklotzde commented:
If we consistently use the class name for the Logger a comma-separated list of class names could be provided for either black or white listing log messages in the MessageHandler.

But before implementing something sophisticated on our own we should look for available logging frameworks that play nicely with Qt.
